lib/web_ui/lib/src/ui/path_metrics.dart

```diff
@@ -120,12 +142,375 @@ class PathMetric {
+    _buildSegments();
+    return true;
+  }
+
+  void _buildSegments() {
+    _segments = <_PathSegment>[];
+    _isClosed = false;
+    double distance = 0.0;
+    int pointIndex = -1;
+    bool haveSeenMoveTo = false;
+    bool haveSeenClose = false;
+
+    if (_path.subpaths.isEmpty) {
+      _contourLength = 0;
+      return;
+    }
+    final engine.Subpath subpath = _path.subpaths[_subPathIndex];
+    final List<engine.PathCommand> commands = subpath.commands;
+    double currentX = 0.0, currentY = 0.0;
+    final Function lineToHandler = (double x, double y) {
+      final double dx = currentX - x;
+      final double dy = currentY - y;
+      final double prevDistance = distance;
+      distance += math.sqrt(dx * dx + dy * dy);
+      // As we accumulate distance, we have to check that the result of +=
+      // actually made it larger, since a very small delta might be > 0, but
+      // still have no effect on distance (if distance >>> delta).
+      if (distance > prevDistance) {
+        _segments.add(_PathSegment(engine.PathCommandTypes.lineTo, distance,
+            [currentX, currentY, x, y]));
+      }
+      currentX = x;
+      currentY = y;
+    };
+    for (engine.PathCommand command in commands) {
+      switch (command.type) {
+        case engine.PathCommandTypes.moveTo:
+          final engine.MoveTo moveTo = command;
+          currentX = moveTo.x;
+          currentY = moveTo.y;
+          haveSeenMoveTo = true;
+          break;
+        case engine.PathCommandTypes.lineTo:
+          assert(haveSeenMoveTo);
+          final engine.LineTo lineTo = command;
+          lineToHandler(lineTo.x, lineTo.y);
+          break;
+        case engine.PathCommandTypes.bezierCurveTo:
+          assert(haveSeenMoveTo);
+          final engine.BezierCurveTo curve = command;
+          // Compute cubic curve distance.
+          distance = _computeCubicSegments(
+              currentX,
+              currentY,
+              curve.x1,
+              curve.y1,
+              curve.x2,
+              curve.y2,
+              curve.x3,
+              curve.y3,
+              distance,
+              0,
+              _kMaxTValue);
+          break;
+        case engine.PathCommandTypes.quadraticCurveTo:
+          assert(haveSeenMoveTo);
+          final engine.QuadraticCurveTo quadraticCurveTo = command;
+          // Compute quad curve distance.
+          distance = _computeQuadSegments(
+              currentX,
+              currentY,
+              quadraticCurveTo.x1,
+              quadraticCurveTo.y1,
+              quadraticCurveTo.x2,
+              quadraticCurveTo.y2,
+              distance,
+              0,
+              _kMaxTValue);
+          break;
+        case engine.PathCommandTypes.close:
+          haveSeenClose = true;
+          break;
+        case engine.PathCommandTypes.ellipse:
+          final engine.Ellipse ellipse = command;
+          _EllipseSegmentResult result = _computeEllipseSegments(
```

Comment:
Again I might be over-optimizing, but since all this is local `_computeEllipseSegments` could take a pre-cached instance of `_EllipseSegmentResult` and populate it with new data rather than instantiating a new object. Flutter uses a similar approach for caching intermediate matrices.

Reply:
Done.